libnmstate/state.py

```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, current):
+        """
+        Merge routes from current to self.
+        If any changed route referring non-exist interface, create the
+        interface state with name only, in order to trigger the recreating of
+        the interface profile.
+        """
+        # Merge other_routes
+        iface_route_sets = defaultdict(set)
+        current_iface_route_sets = defaultdict(set)
+
+        for route in self.routes:
+            if Route.NEXT_HOP_INTERFACE in route:
+                iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                    _Route(route))
+            else:
+                logging.warning("Ignoring the route entry with no "
+                                "next hop interface defined: %s",
+                                route)
+
+        for route in current.routes:
+            iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                _Route(route))
+            current_iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                _Route(route))
+
+        for iface_name in list(six.viewkeys(iface_route_sets)):
```

Comment:
What is the `list` all about?

Reply:
Since the `iface_route_sets` might change at `del iface_route_sets[iface_name]`, using `list()` could present the warning about change when iterating.

Reply:
This is why you should not remove entries in the middle of the loop. Use `filter` or a an immutable approach to do it.